language/move-compiler/src/typing/translate.rs

```diff
@@ -523,16 +533,14 @@ fn struct_def(context: &mut Context, s: &mut N::StructDefinition) {
     for (_field_loc, _field, idx_ty) in field_map.iter() {
         let loc = idx_ty.1.loc;
         let subst_ty = core::subst_tparams(tparam_subst, idx_ty.1.clone());
-        let inst_ty = core::instantiate(context, subst_ty);
```

Comment:
The root of the bug is that `instantiate` adds the ability constraints to the context. But, for abilities we substitute in the Any type before checking ability field requirements. So the type parameters would erroneously satisfy the constraints even when they shouldn't.

The fix is then to instantiate the type for the constraints separately from the Any substitution.